src/field_theory/adjoin.lean

```diff
@@ -0,0 +1,177 @@
+    one_mem' := is_submonoid.one_mem,
+    mul_mem' := λ x y, is_submonoid.mul_mem,
+    zero_mem' := is_add_submonoid.zero_mem,
+    add_mem' := λ x y, is_add_submonoid.add_mem,
+  },
+  algebra_map_mem' := λ x, field.mem_closure (or.inl (set.mem_range.mpr ⟨x,rfl⟩)),
+}
+
+lemma adjoin.field_mem (x : F) : algebra_map F E x ∈ adjoin F S :=
+field.mem_closure (or.inl (set.mem_range_self x))
+
+lemma adjoin.field_subset : set.range (algebra_map F E) ⊆ adjoin F S :=
+begin
+  intros x hx,
+  cases hx with f hf,
+  rw ←hf,
+  exact adjoin.field_mem F S f,
+end
+
+instance adjoin.field_coe : has_coe_t F (adjoin F S) :=
+{coe := λ x, ⟨algebra_map F E x, adjoin.field_mem F S x⟩}
+
+lemma adjoin.set_mem (x : S) : ↑x ∈ adjoin F S :=
+field.mem_closure (or.inr (subtype.mem x))
+
+lemma adjoin.set_subset : S ⊆ adjoin F S :=
+λ x hx, adjoin.set_mem F S ⟨x,hx⟩
+
+instance adjoin.set_coe : has_coe_t S (adjoin F S) :=
+{coe := λ x, ⟨↑x, adjoin.set_mem F S x⟩}
+
+lemma adjoin.mono (T : set E) (h : S ⊆ T) : (adjoin F S : set E) ⊆ adjoin F T :=
+field.closure_mono (set.union_subset (set.subset_union_left _ _) (set.subset_union_of_subset_right h _))
+
+instance adjoin.is_subfield : is_subfield (adjoin F S : set E) := field.closure.is_subfield
+
+lemma adjoin_contains_field_as_subfield (F : set E) {HF : is_subfield F} : F ⊆ adjoin F S :=
+λ x hx, adjoin.field_mem F S ⟨x, hx⟩
+
+lemma adjoin_contains_subset {T : set E} {H : T ⊆ S} : T ⊆ adjoin F S :=
+begin
+  intros x hx,
+  exact adjoin.set_mem F S ⟨x,H hx⟩,
+end
+
+/-- proves an obvious linear equivalence (occasionally useful when working with dimension) -/
+definition adjoin_as_submodule_equiv : (adjoin F S) ≃ₗ[F] (adjoin F S) := {
+  to_fun := λ x, x,
+  map_add' := λ x y, rfl,
+  map_smul' :=
+  begin
+    intros x y,
+    ext1,
+    change _ = x • ↑y,
+    rw algebra.smul_def,
+    rw algebra.smul_def,
+    refl,
+  end,
+  inv_fun := λ x, x,
+  left_inv := λ x, rfl,
+  right_inv := λ x, rfl,
+}
+
+/-- If F ⊆ T and S ⊆ T then F[S] ⊆ F[T] -/
+lemma adjoin_subset {T : set E} [is_subfield T] (HF : set.range (algebra_map F E) ⊆ T)
+(HS : S ⊆ T) : (adjoin F S : set E) ⊆ T :=
+begin
+  apply field.closure_subset,
+  rw set.union_subset_iff,
+  exact ⟨HF,HS⟩,
+end
+
+/-- If S ⊆ F[T] then F[S] ⊆ F[T] -/
+lemma adjoin_subset' {T : set E} (HT : S ⊆ adjoin F T) : (adjoin F S : set E) ⊆ adjoin F T :=
+adjoin_subset F S (adjoin.field_subset F T) HT
+
+lemma set_range_subset {T₁ T₂ : set E} [is_subfield T₁] {hyp : T₁ ⊆ T₂} :
+set.range (algebra_map T₁ E) ⊆ T₂ :=
+begin
+  intros x hx,
+  cases hx with f hf,
+  rw ←hf,
+  cases f with t ht,
+  exact hyp ht,
+end
+
+lemma adjoin_contains_field_subset {F : set E} {HF : is_subfield F} {T : set E} {HT : T ⊆ F} :
+T ⊆ adjoin F S :=
+λ x hx, adjoin.field_mem F S ⟨x,HT hx⟩
+
+/-- F[S][T] = F[S ∪ T] -/
+lemma adjoin_twice (T : set E) : (adjoin (adjoin F S : set E) T : set E) = adjoin F (S ∪ T) :=
+begin
+  apply set.eq_of_subset_of_subset,
```

Comment:
If you have two goals, please use `{ .. }` and indentation to reflect this in your proof.